SWIPs/swip-honey_to_money.md

```diff
@@ -7,51 +7,66 @@ Type: Standards track
-
-In the current implementation, honey prices are hardcoded. When this price in the codebase changes, it does not affect nodes running a previous release; an old release expects a settlement of honey to be worth a different amount in currency. Eventually, this will result in nodes who run different versions of the software to disconnect from each other as the difference in prices will cause balances to be updated differently on each node. This SWIP proposes an on-chain price oracle to enable an atomic update process of honey prices without requiring nodes to update their source-code.
+This update process should be atomic: either all nodes upgrade or none. If Swarm would update the price by releasing a new version of the source code (as is currently the case), this would be violated as it is not possible to guarantee that all nodes are on the same release. By externalizing the price to an oracle, Swarm completely decouples its core functions from price discovery. By building support for price oracles, it will be possible to coordinate price changes across the network and set the stage for future experimentation with various mechanisms for setting prices, from DAOs, to bonding curves, to fixed prices. 
 
 ## Specification
-* The Swarm source code will reference the address of a `HoneyToMoney` price oracle
-* The price oracle implements the `HoneyToMoney` interface (to-be specified) 
-* The `HoneyToMoney` contract will be initially owned by an M/N multi-sig of Swarm developers and stakeholders.
-* Before sending a payment (denominated in currency), a node will query the `HoneyToMoney` price oracle to get the most up-to-date price conversion. 
-* Upon receiving a payment, nodes will determine the honey amount paid by looking up and dividing by the currently valid price from the oracle. This might cause payment imbalances between nodes, as the oracle might be queried one or several blocks before the transaction was sent. To avoid this situation, we propose that nodes include a payload with the transaction, specifying the time at which the oracle was queried.
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for the current Swarm platform and future client implementations.-->
+* An oracle is reachable at one specific endpoint and this endpoint is hardcoded in the Swarm source-code. 
```

Comment:
Is that correct? I thought it is not hardcoded, but supplied via parameter?

Reply:
wouldn't it be a problem to have each node consult its own oracle?

Reply:
i think this is referenced in the _Abstract_ section:
>this SWIP includes a proposal on implementation of an *on-chain* price oracle, **governed by Swarm developers and stakeholders** to allow for experimentation.

Reply:
No, I mean the actual oracle address. This has to be a reachable something. Thus, either it is indeed hardcoded (which means if we change the oracle we need to do a release, and I would vote against that), or it is somehow supplied. I think it should be a command line parameter, but of course, the actual fate of the mutli-payment SWIP is related to this one, because I guess it would be supplied via the parameters described there.

Reply:
i agree that changing an oracle address shouldn't mean releasing a new version :+1:

Reply:
Two nodes who refer to a different oracle cannot engage in *SWAP* with each other, as there is no guarantee that the oracles will quote the same price (resulting in possible accounting differences). I think we can best compare it to a parameter such as the allowedNetworkID, this one is hardcoded with a default value, but can be changed via a config file or flag if required (implicit via swap-backend-url).  In a new release, a different default value can be defined if needed. 

The multiple-payments SWIP might mean a change in implementation for this, but the logic will stay the same: a payment module references an oracle as a default, but nodes can change this if they so desire. 

@holisticode, do you require a change here? I personally would vote for no, as this is the specification of what is minimally needed and, while the option to provide it yourself via config or flag is nice, I would say that it is not minimally needed (unless you want to *only* provide it via config/flag, which, I think would be undesirable since we this will make it more likely that nodes are going to provide different oracles).

Reply:
I think I would like a change yes. If the oracle is hardcoded, how are we going to test? Should we not be able to provide different oracles depending on where "we are" (test environment, production, etc.)?

What if the contract of the oracle has an issue and needs to be changed? Then we need to do a release?

Yes, I think it should not be hardcoded and should be provided as a parameter.

Reply:
> What if the contract of the oracle has an issue and needs to be changed? Then we need to do a release?

If the oracle address can be provided as a parameter, there is no need to do a new release. However, to update the default a new release is indeed needed. I agree that it is a good idea to be able to supply the address of the oracle via config/flag in order to do testing, but also to facilitate nodes to update the oracle address in case of emergency (Swarm devs going rogue, hack, vulnerability in smart-contract, etc).
 
Will change!

Reply:
Changed into:
```
* An oracle is reachable at one specific endpoint and this endpoint is hardcoded in the Swarm source-code as a default parameter, but can be adjusted by the node if so desired (e.g. via config and/or flag).
```

Reply:
The main reason why it should be hardcoded is that nodes are not guaranteed to apply the same prices if they are not listening to the same oracle. In case we would not hardcode it, but only let it be configurable, nodes will need to go online, search for which oracle to use, pass it in (either config/flag): this is bad UX. As 99% of the nodes won't like to configure this parameter, I think it would be best to keep the text as-is.

As we agreed on: we are merging this SWIP now. Feel free to raise this as a point of discussion with the wider community: I'll be gladly proven wrong in my assumption!